### ISSUE TITLE
[Modern] "Dumb" QueryRenderer equivalent

### ISSUE THREAD

**Issue by taion** [facebook/relay#1687]
I'm looking to build an equivalent to react-router-relay or Found Relay for use with Relay Modern, to handle fetching multiple "root" queries in parallel, despite those queries corresponding to nested routes and corresponding components.

I don't think the approach I've taken for either library quite works here. They're respectively:
- Aggregate a bunch of queries configs into a single query config, but Relay Modern doesn't have query configs
- Use `primeCache`, but Relay Modern doesn't do caching in all cases

The approach I plan to explore is pull out the logic up to `environment.streamQuery` and `environment.subscribe` in `<QueryRenderer>` into a separate function, and just hand off the snapshot to something like a "dumb query renderer". This seems analogous to the divide between `<Relay.Renderer>` and `<Relay.ReadyStateRenderer>` in Relay Classic.

The next-best option I could think of was to render a bunch of `<QueryRenderer>`s in parallel at the root level, then portal through the data via `context` or something, but that seemed pretty gross. Also, pulling the initial snapshot logic out of React components makes for a nicer server rendering API.

Given that, does the approach I've outlined above sound reasonable? If so, would it make sense for this "dumb query renderer" to live in Relay core? I believe it'd be something that could be factored out of the current `<QueryRenderer>`, and that equivalent split between `<Relay.Renderer>` and `<Relay.ReadyStateRenderer>` did exist in Relay Classic.

**Comment by josephsavona** [comment]
Great question! I can imagine a bunch of different ways to go here. First some context, then suggestions.

> Aggregate a bunch of queries configs into a single query config, but Relay Modern doesn't have query configs

Query configs (routes, really) were primarily necessary in Relay classic to work around the fact that legacy, FB-internal GraphQL didn't support multiple root fields. Relay Modern is built around the OSS GraphQL spec, so query configs are replaced with just queries and variables. Static queries also means that it isn't really possible to aggregate multiple queries together beyond possibly batching multiple queries into a single HTTP request at the network layer, if your server supports unwrapping them and execute multiple queries.

> Use primeCache, but Relay Modern doesn't do caching in all cases

Relay Modern doesn't do caching at all *by default*. However all the tools are there. The main approaches are request/response caching at the network layer (we actually have a RelayQueryResponseCache class that should probably be exported) and to add new QueryRenderer-style APIs that use `lookup` and friends to try reading from local data first.

Routing and server rendering are a great examples of the types of functionality we wanted to enable by making the `react-relay` and `relay-runtime` separate modules. Long term it may make sense to incorporate some of your suggestions into QueryRenderer itself and split it along the lines of the classic RelayRenderer/ReadyStateRenderer. As a first step though, I'd encourage you to write this as a standalone API that uses functions from `relay-runtime`, perhaps based on `QueryRenderer`, share it here, and then we can discuss next steps!

This would be a great test to make sure that the runtime is documented (you'll have to look at the source, I suggest RelayStaticEnvironment, RelayCore, and the architecture docs) and provides all the necessary features.

**Comment by taion** [comment]
Right I'm just going to copy-paste bits from `<QueryRenderer>` first. I think that's all the code I need. I don't need to introduce any concept of caching, I don't think – I just need to pass down the `snapshot` to something like a `<QuerySnapshotRenderer>`, unless I'm missing something.

**Comment by robrichard** [comment]
I'm trying to write something very similar for server side rendering. I forked `QueryRenderer` to create a similar component that does `environment.lookup` to see if the data is already in the environment before rendering. I was able to get this to work for the [server render](https://github.com/robrichard/relay-examples/blob/ssr-example/todo-modern-server-side-rendering/js/renderServer.js#L19-L43) and for the [initial client render](https://github.com/robrichard/relay-examples/blob/ssr-example/todo-modern-server-side-rendering/js/app.js#L24-L40), but the component does not re-render when the store is updated on mutation.

This is what I added to `QueryRenderer`: https://github.com/robrichard/relay-examples/blob/ssr-example/todo-modern-server-side-rendering/js/components/ReactRelayLookupRenderer.js#L48-L63

**Comment by josephsavona** [comment]
> I just need to pass down the snapshot to something like a <QuerySnapshotRenderer>, unless I'm missing something.

@taion Yeah that should work

**Comment by josephsavona** [comment]
@robrichard Hmm that looks like it should work. I would check with `componentWillReceiveProps()` or some other lifecycle hook is perhaps getting executed and wiping out the subscription? Also, `lookup(selector)` will always return a snapshot for which the `data` property is non-null so your conditional may not be entering the right branch. `lookup()` returns whatever happens to be in the store; the fact that it returns a non-null result doesn't mean that exact query has been fetched before. The strategy @taion described is probably more robust: explicitly fetch the data, call lookup() and then pass a snapshot.

**Comment by robrichard** [comment]
So a minimal snapshot renderer would look like this? Before rendering you would need to manually fetch the query from the environment, lookup the snapshot, and subscribe to updates if necessary? Is there anything missing here?

```js
class RelaySnapshotRenderer extends Component {
    render() {
        const {snapshot, render} = this.props;
        return render({props: snapshot.data});
    }
    getChildContext() {
        return {
            relay: {
                environment: this.props.environment,
                variables: this.props.snapshot.variables
            }
        };
    }
}

RelaySnapshotRenderer.childContextTypes = {
    relay: PropTypes.object.isRequired
};
```

**Comment by taion** [comment]
I think you'd want to handle calling `environment.subscribe` as well, at least on the client. I'm really looking at the code in `<QueryRenderer>` that happens logically after `snapshot` is first bound.

**Comment by josephsavona** [comment]
cc @taion re 1c585b2

**Comment by taion** [comment]
I guess if we want something like the old `forceFetch` behavior where an initial render shows the stale data, but then a subsequent re-render shows updated data once available, we'll have to code that up ourselves?

**Comment by josephsavona** [comment]
@taion Yup! With `check` that's now possible:

```
if (check()) {
  // lookup() and render
}
fetch() // and then re-lookup and render when complete
```

**Comment by robrichard** [comment]
For anyone else that comes across this issue, this is the component I'm using now. @taion @josephsavona let me know if you see any issues with this approach.

```js
import {Component} from 'react';
import PropTypes from 'prop-types';

class RelaySnapshotRenderer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            data: props.snapshot.data
        };
        this.onChange = this.onChange.bind(this);
    }
    getChildContext() {
        return {
            relay: {
                environment: this.props.environment,
                variables: this.props.snapshot.variables
            }
        };
    }
    componentDidMount() {
        const {snapshot, environment} = this.props;
        this.subscription = environment.subscribe(snapshot, this.onChange);
    }
    componentWillUnmount() {
        this.subscription.dispose();
        this.subscription = null;
    }
    onChange(snapshot) {
        this.setState({
            data: snapshot.data
        });
    }
    render() {
        const {render} = this.props;
        const {data} = this.state;
        return render({props: data});
    }
}

RelaySnapshotRenderer.childContextTypes = {
    relay: PropTypes.object.isRequired
};

RelaySnapshotRenderer.propTypes = {
    environment: PropTypes.object.isRequired,
    snapshot: PropTypes.object.isRequired,
    render: PropTypes.func.isRequired
};

export default RelaySnapshotRenderer;
```

**Comment by josephsavona** [comment]
@robrichard That looks about right. One very important missing detail: the component needs to implement `componentWillReceiveProps`, discarding the previous subscription if either the snapshot or environment prop changes.

Given that this component itself subscribers for changes, it might make more sense for this to component to accept a `selector: Selector` prop instead of a snapshot. As is, the type signature allows the parent to also observe the value, whereas a selector makes it clear that the component itself will do the subscription.

**Comment by robrichard** [comment]
@josephsavona does it make sense to add a `lookup={true}` prop on `QueryRenderer` to run `check()` and `lookup()` before attempting to fetch data? It seems otherwise we'll need to re-implement a lot of the logic that's currently in `QueryRenderer`, to properly handle subscriptions and edge cases regarding environment and variables changing?

**Comment by robrichard** [comment]
@josephsavona I opened a PR (https://github.com/facebook/relay/issues/1760) to support running `check` and `lookup` in `QueryRenderer`. Let me know whether or not you think this is the right approach.

@taion would this also solve your use case?

**Comment by taion** [comment]
Potentially – I was thinking of passing the initial snapshot into `QueryRenderer` directly, but quite possibly this approach is better.

**Comment by unirey** [comment]
@josephsavona  `RelayQueryResponseCache` looks very interesting because now when we have static queries, request/response level caching seems appropriate and very cheap to check against. I also love the fact that `ttl` is already implemented and that it uses `queryID` which plays very well with persistent queries.

It has `ttl` that applies to every item in that cache instance. Intentional? Create multiple instances if different queries have different `ttl`? Another idea would be to add `ttl` as a setter argument and store it with every item separately, use instance `ttl` only as a default. But I guess it's easy to implement if anyone needs that level of flexibility because it does have minor overhead (questionable levels in most apps, but still..).


This is roughly how it's meant to be used?


```js
import RelayQueryResponseCache from 'relay-runtime/lib/RelayQueryResponseCache.js';
const cache = new RelayQueryResponseCache({size: 250, ttl: 60 * 5 * 1000 });

// funtion used in Network.create(fetchQuery);
function fetchQuery(operation, variables, config, files) {
  const queryID = /* not sure if & how Relay helps w/ persistent queries (gen ID for queries) */;

  const data = cache.get(queryID, variables);
  if (data !== null) return data; // cache hit or miss

  return fetch(GRAPHQL_ENDPOINT, {
    method: 'POST',
    headers: {
      'authorization': 'auth-token-123',
      'content-type': 'application/json'
    },
    body: JSON.stringify({
      queryID: queryID, // send only ID, queries also stored server-side
      variables: variables
    })
  })
  .then(res => {
    return res.json();
  })
  .then(jsonPayload => {
    if (jsonPayload.data) {
      cache.set(queryID, variables, jsonPayload.data);
      return jsonPayload;
    }

    // errors etc
  });
}
```

**Comment by unirey** [comment]
@josephsavona  `RelayQueryResponseCache` looks very interesting because now when we have static queries, request/response level caching seems appropriate and very cheap to check against. I also love the fact that `ttl` is already implemented and that it uses `queryID` which plays very well with persistent queries.

It has `ttl` that applies to every item in that cache instance. Intentional? Create multiple instances if different queries have different `ttl`? Another idea would be to add `ttl` as a setter argument and store it with every item separately, use instance `ttl` only as a default. But I guess it's easy to implement if anyone needs that level of flexibility because it does have minor overhead (questionable levels in most apps, but still..).


This is roughly how it's meant to be used?


```js
import RelayQueryResponseCache from 'relay-runtime/lib/RelayQueryResponseCache.js';
const cache = new RelayQueryResponseCache({size: 250, ttl: 60 * 5 * 1000 });

// function used in Network.create(fetchQuery);
function fetchQuery(operation, variables, config, files) {
  const queryID = /* not sure if & how Relay helps w/ persistent queries (gen ID for queries) */;

  const data = cache.get(queryID, variables);
  if (data !== null) return data; // cache hit

  // cache missed
  return fetch(GRAPHQL_ENDPOINT, {
    method: 'POST',
    headers: {
      'authorization': 'auth-token-123',
      'content-type': 'application/json'
    },
    body: JSON.stringify({
      queryID: queryID, // send only ID, queries also stored server-side
      variables: variables
    })
  })
  .then(res => {
    return res.json();
  })
  .then(jsonPayload => {
    if (jsonPayload.data) {
      cache.set(queryID, variables, jsonPayload.data);
      return jsonPayload;
    }

    // errors etc
  });
}
```

**Comment by josephsavona** [comment]
Yup, that's basically how it's meant to be used! One thing to note is that you should clear the cache whenever a mutation is executed since mutations could invalidate cached results (check the type/kind property of the operation object, can't remember offhand which). 

As for the global TTL: the cache-wide TTL setting is for simplicity. You could certainly build your own cache and do something else. I think the answer long-term is fragment level TTL: a query is only reexecuted if any fragment data is missing/stale. For now, global TTL is a quick way to ensure you don't re-execute a query twice in a short time interval. 

There's a separate issue open for generating persisted queries, on mobile so can't find the link just now.

**Comment by unirey** [comment]
@josephsavona I regress, mutation-heavy apps are not taking much advantage of `RelayQueryResponseCache`, it has too robust invalidation, I prefer something more granular.

Im trying to make my own `QueryRenderer` implementation with cache and serving partial data from cache, something like this: 


```js
const hasAllData = environment.check(operation.root);
const snapshot = environment.lookup(operation.fragment);

if (!hasAllData) {
  // if has ANY data (might require more complex comparison, haven't dig into it yet)
  if (snapshot.data !== undefined) { 
    // render UI with partial data (e.g above to fold, React components handle missing data)
    readyState = {error: null, props: snapshot.data, retry: null};
  } else {
    // no cached data at all, render spinner, shell or something else
    readyState = {error: null, props: null, retry: null};
  }

  fetch() // and then re-lookup and re-render with 100% fulfilled data
} else {
  // has everything -> render
  readyState = {error: null, props: snapshot.data, retry: null};
}
```

Then Im able to use the data I already have in cache to render above the fold (or anything really).

Most common use case: 
* User scrolls through the list (have some data for each item, could be anything: title, author, avatar _etc_)
* If user clicks on an item, route to single item view and it gets queried with `node` interface
* Already have basic data about the item, render it, virtually instant "time to first meaningful paint"
* Rest of the data arrives and often gets rendered below the fold even before user starts scrolling

Do you see anything obvious that might burst my bubble?

**Comment by josephsavona** [comment]
@unirey The basic approach you're describing for an alternate take on QueryRenderer makes sense. Their is one potential blocker, though: GraphQL does not allow clients to understand that `query Foo { node(id: "xyz") { ... } }` means "start from record xyz". So if you navigate from a list view (and have fetched record xyz as one of the items) to the detail view, Relay (and any other client) will see a root field that it has not yet fetched (`node(id: xyz)`) and return undefined. This is something we'd like to address long-term, but there isn't a good solution for it today.

Note that Relay Classic treated the `node` root field specially; in Relay Modern we've tried to make as few assumptions as possible about the schema and not have magic behavior for certain fields/arguments.

**Comment by unirey** [comment]
@josephsavona Ah, I see. That's little disappointing I must admit but there are bigger thing to worry about.

---

@robrichard I can't wait until PR #1760 (or something similar) gets added to Relay but one problem still stands: if you render separate `QueryRenderer` for each route/component (which is the easiest solution in lots of cases), cached data is disposed if `QueryRenderer` unmounts. At the moment only way to keep the cache seems to be to use single `QueryRenderer` and only swap out the `query` prop (haven't tried yet). 

@josephsavona Would love to know a little bit more about garbage collection and how it correlates to `retain` and `dispose`. My current assumptions:
* data that is not retained, gets garbage collected (and we can't control it, at least not yet)
* data that is retained, never gets garbage collected until `dispose` is called

Data seems to be deleted as soon as `QueryRenderer` is unmounted (which calls `dispose`), does that mean that `dispose` sends a signal to get rid of it ASAP?

**Comment by taion** [comment]
Why would the `<QueryRenderer>` get unmounted entirely? Or rather, in your routing tree, the `<QueryRenderer>` that interacts with the same data should stay at similar positions, no? I'm thinking something like your navigation/sidebar elements that don't really move around in your component tree much.

**Comment by josephsavona** [comment]
Re GC:
- calling retain() ensures the data for thatbwuery won't be disposed
- dispose()-ing the retain will cause GC to be scheduled (so that if you dispose multiple values in the same event loop GC runs only once). It will clear everything that is not still retained.

**Comment by unirey** [comment]
@taion Something like this? Im not sure how else to combine routing and Relay Modern.

```jsx
<Router>
  <Route path="/profile" component={FirstPageWrapper}/>
  <Route path="/gallery" component={OtherPageWrapper}/>
</Router>

<FirstPageWrapper><QueryRenderer/></FirstPageWrapper>
<OtherPageWrapper><QueryRenderer/></OtherPageWrapper>
```

Wrapper components take care of query, variables, rendering logic _etc_. My quick and naive approach before better arhitecture or integration with routing libraries surfaces. Would love to hear better ideas.

**Comment by taion** [comment]
@unirey 

No, it will be more or less exactly what's described here – prefetch the snapshots, then inject into dumb/lookup-based query renderers.

I'm just starting with the code now, but it should look virtually the same as react-router-relay or Found Relay, and internally will work pretty similar to how Found Relay works.

**Comment by unirey** [comment]
@taion What exactly do you mean with "prefetch the snapshots" and do you plan to use one root level renderer or per route? Sorry about all the probably stupid questions but I never got into Relay Classic source code and Im unable to draw any parallels and I need to solve routing quickly, even if it's dirty.

**Comment by taion** [comment]
I ended up moving almost all of the logic out of my "dumb" `<QueryRenderer>` equivalent to make it simpler to manage the lifecycle for multiple routes.

I ended up with just: https://github.com/4Catalyzer/found-relay/blob/v0.3.0-alpha.1/src/modern/ReadyStateRenderer.js.

**Comment by taion** [comment]
Well, the thing I ended up with is so simple that I don't really need anything here. I'm going to close this out for now.

**Comment by IkerArb** [comment]
@josephsavona how and where do I use the retain() method, I am kind of new here and I am using React Native with independent QueryRenderers, when I pop a view the previous caché is being emptied, I want to know how can I retain my cache

**Comment by josephsavona** [comment]
@IkerArb can you file a new issue and describe the issue further?

**Comment by tomas-paronai** [comment]
I got a bit lost here. Not sure now how should we put a simple caching in place? 
@unirey s [comment](https://github.com/facebook/relay/issues/1687#issuecomment-302931855) was interesting, but how do we achieve to have persistence then?

**Comment by jacobpdq** [comment]
I've decided to render my graphql responses to static JSON files & serve those locally. 

Hope this helps.

**Comment by ErrorPro** [comment]
I decided to use `operation.hash` (it is being generated by babel-relay-plugin and it is stored inside __generated__ files) so it only changes whenever you change a query/fragment. I'm not sure how to reset cache when a mutation is committed(haven't got there yet) but for caching requests it's working so far. What do you think @josephsavona ?

**Comment by alloy** [comment]
@ErrorPro Makes sense to me 👍

**Comment by taion** [comment]
You need both the hash and the variables, BTW.

**Comment by ErrorPro** [comment]
@taion https://github.com/facebook/relay/blob/master/packages/relay-runtime/network/RelayQueryResponseCache.js#L67
